packages/gatsby/src/utils/page-data.js

```diff
@@ -1,28 +1,54 @@
 const fs = require(`fs-extra`)
 const path = require(`path`)
+const Promise = require(`bluebird`)
+const Worker = require(`jest-worker`).default
+const cpuCoreCount = require(`./cpu-core-count`)
+const { chunk } = require(`lodash`)
 
 const getFilePath = ({ publicDir }, pagePath) => {
   const fixedPagePath = pagePath === `/` ? `index` : pagePath
   return path.join(publicDir, `page-data`, fixedPagePath, `page-data.json`)
 }
-
 const read = async ({ publicDir }, pagePath) => {
   const filePath = getFilePath({ publicDir }, pagePath)
   const rawPageData = await fs.readFile(filePath, `utf-8`)
   return JSON.parse(rawPageData)
 }
 
-const write = async ({ publicDir }, page, result) => {
+const write = async ({ publicDir }, page, result, webpackCompilationHash) => {
   const filePath = getFilePath({ publicDir }, page.path)
   const body = {
     componentChunkName: page.componentChunkName,
     path: page.path,
+    webpackCompilationHash,
     result,
   }
   await fs.outputFile(filePath, JSON.stringify(body))
 }
 
+const updateCompilationHashes = (
+  { publicDir },
+  pagePaths,
+  webpackCompilationHash
+) => {
+  const workerPool = new Worker(require.resolve(`./page-data-worker`), {
```

Comment:
I'm surprised a big that this makes things faster? What was the difference? Node does use a thread pool for FS. Hmmm looking at this https://stackoverflow.com/questions/20346097/does-node-js-use-threads-thread-pool-internally there is limits so for mass updates I can see the limits on fs threads causing trouble.

Reply:
So, I haven't actually run benchmarks on this, but my thinking is that the JSON deserialization/serialization is computationally expensive, so running that work on multiple workers will result in a performance increase. Especially since no data needs to be passed between processes (except the file path). Basically, this is a CPU optimization, not an I/O one.

Reply:
Yeah — seems reasonable — does this reuse the workers for rendering pages? It'd be nice not to spawn extra Node.js processes if we don't have to.

Reply:
No, it's a different worker pool. Interesting idea on sharing them. The challenge is that we'd have to write a new `worker.js` that references the code for both html rendering and this code, and then export the functions for each. That's totally doable, but might add a bit too much complexity for now. I'd say that if we use jest-worker in any other parts of gatsby, it might make sense to consolidate at that point.

Reply:
this sounds a lot like a "job" api that handles things to a worker pool.

Reply:
@Moocar we'd just have to merge the two files right and have different exports for the two? Am I missing something? Doesn't seem too complex.

Reply:
@KyleAMathews Yep, and then make sure we start and end the worker pool at the right place. More than happy to code this up for this PR.

Reply:
Thanks, let's do that — this will make things more efficient + it'll make it really easy to add the worker pool to other tasks as they come up as we'll have the basic infra in place + since the workers are already spun up, sending task to them is really cheap.